### ISSUE TITLE
Intersector.intersectRayBounds can return an intersection outside the bounding box

### ISSUE THREAD

**Issue by fwolff** [libgdx/libgdx#3120]
How to reproduce:

```java
Ray ray = new Ray(new Vector3(5731.2046f, -6039.354f, 11157.498f), new Vector3(-0.62397736f, -0.053491283f, -0.77960944f));
BoundingBox box = new BoundingBox(new Vector3(-3840.0f, -7680.0f, 848.0f), new Vector3(-1920.0f, -5760.0f, 1929.0f));
Vector3 entry = new Vector3();
if (Intersector.intersectRayBounds(ray, box, entry))
    System.out.println(entry); // [-1919.9995, -6695.2637, 1597.9346] (error: -1919.9995 > -1920.0)
```

Suggested patch (Intersector.intersectRayBounds):

```java
if (hit && intersection != null) {
    intersection.set(ray.direction).scl(lowest).add(ray.origin);
    if (intersection.x < box.min.x)
        intersection.x = box.min.x;
    else if (intersection.x > box.max.x)
        intersection.x = box.max.x;
    if (intersection.y < box.min.y)
        intersection.y = box.min.y;
    else if (intersection.y > box.max.y)
        intersection.y = box.max.y;
    if (intersection.z < box.min.z)
        intersection.z = box.min.z;
    else if (intersection.z > box.max.z)
        intersection.z = box.max.z;
}
```

Additionally, if the ray origin is inside the box, the code doesn't set the intersection point:

```java
if (v0.x > 0 && v0.y > 0 && v0.z > 0 && v1.x < 0 && v1.y < 0 && v1.z < 0) {
    return true;
}
```

I think the intersection point should be set to the ray origin in this case, even if it can break existing code... Suggested patch:

```java
if (v0.x > 0 && v0.y > 0 && v0.z > 0 && v1.x < 0 && v1.y < 0 && v1.z < 0) {
    if (intersection != null)
        intersection.set(ray.origin); 
    return true;
}
```

Thanks!

**Comment by xoppa** [comment]
Thanks! Looks like a good addition, would you mind to [send a pull request](https://github.com/libgdx/libgdx/wiki/Contributing) for it?

If there's an intersection and the intersection point is requested then that point should be within the bounds of the bounding box. Clamping the value to avoid small floating point errors sounds like a good way to guarantee that.

If the origin is inside the box then indeed the intersection point should be set as well. Setting it to the origin looks like the most obvious and simple solution for that. In that case the intersection point would be defined as "the point on the ray closest to the origin which is within the specified bounds".

Someone might expect the intersection point to be both *on* the ray and *on* the bounding box (each of the six planes) though. Floating point errors are to be expected, but having an intersection point *inside* the bounding box might not be that obvious. So adding a line to the javadocs might be a good idea.

**Comment by fwolff** [comment]
Pull request sent.

I have added a comment about what is returned by this method, in particular in the case of a ray origin inside the box.

Regarding this part of your comment: 

> Someone might expect the intersection point to be both *on* the ray and *on* the bounding box (each of the six planes) though.

I think this is not the only issue with my fix: the way I'm clamping the intersection point can make it slightly diverge from the ray. I don't think this is problem (at least it is not an issue for me), but I say it explicitly in the javadoc to dissipate any ambiguity or wrong expectation.

Anyway, many thanks for this awesome project and your kind response!

**Comment by fwolff** [comment]
I just realized there is another issue with this method (even with my patch). The following code is wrong:

```java
v0.set(ray.origin).sub(box.min);
v1.set(ray.origin).sub(box.max);
if (v0.x > 0 && v0.y > 0 && v0.z > 0 && v1.x < 0 && v1.y < 0 && v1.z < 0) {
	return true;
}
```

This is obvious if you set the origin of box to (say) the minimum of the box, but it is also true if it is on any of the six planes of the box: the code tests if the origin is *exclusively* inside the box (excluding the bounds). The simplest fix is:

```java
if (box.contains(ray.origin)) {
	if (intersection != null) intersection.set(ray.origin);
	return true;
}
```

I'm preparing a new pull request, but I want first to fix another issue I found with the intersectRayTriangle method (see next issue).
